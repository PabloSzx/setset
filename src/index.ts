export * from './manager'
export * from './settings'
export * from './static'
