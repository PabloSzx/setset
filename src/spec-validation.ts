import * as Lo from 'lodash'
import { inspect } from 'util'
import {
  appendPath,
  isNamespaceSpecifier,
  isRecordSpecifier,
  renderPath,
  Specifier,
  SpecifierNamespace,
  TraversalInfo,
} from './settings'

/**
 * Validate the spec for basic invariants.
 */
export function validateSpecifier(specifier: Specifier, info: TraversalInfo) {
  if (isNamespaceSpecifier(specifier)) {
    if (specifier.map !== undefined && typeof specifier.map !== 'function') {
      let message = ''
      if (info.path.length === 1) message += 'Root mapper was invalid.'
      else message += `Mapper for setting specifier at path "${renderPath(info)}" was invalid.`
      // prettier-ignore
      message  += ` Mappers must be functions. Got: ${inspect(specifier.map)}`
      throw new Error(message)
    }

    Lo.forOwn(specifier.fields, (fieldSpecifier: Specifier, fieldName: string) => {
      validateSpecifier(fieldSpecifier, appendPath(info, fieldName))
    })

    return
  }

  if (isRecordSpecifier(specifier)) {
    validateSpecifier(specifier.entry as SpecifierNamespace, info)
    return
  }
}
