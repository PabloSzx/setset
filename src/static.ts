/**
 * This module only deals with the type level (no term level)
 * It contains the rich conditional type system powering the settings library.
 */

import { Primitive } from 'type-fest'
import {
  ExcludeUndefined,
  IncludesPlainObjectOrInterface,
  IncludesRecord,
  IsEqual,
  IsRecord,
  KeepOptionalKeys,
  KeepRequiredKeys,
  Lookup,
  OnlyPlainObjectOrInterface,
  PlainObject,
} from './utils'

// todo implement context
type BaseContext = { path: string[] }

export type Spec<Input, Data> = NamespaceSpec<Input, Data, {}>

export type InferDataFromInput<Input> = {
  [K in keyof Input]-?: Input[K] extends Leaf
    ? ExcludeUndefined<Input[K]['type']>
    : IncludesPlainObjectOrInterface<Input[K]> extends true
    ? InferDataFromInput<OnlyPlainObjectOrInterface<Input[K]>>
    : ExcludeUndefined<Input[K]>
}

export type UserInput<Input> = UnwrapSyntheticLeavesDeeply<Input, true>

export type KeysWhereNotPresentOrRequiredInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? never : K) : K
}[keyof Input]

type KeysWhereNotPresentOrOptionalInData<Input, Data> = {
  [K in keyof Input]: K extends keyof Data ? (undefined extends Data[K] ? K : never) : K
}[keyof Input]

type NO_DATA_MATCH = 'NO_DATA_MATCH'

type Node<Input, Data, Key, LocalContext> = Lookup<Input, Key> extends Leaf
  ? LeafSpec<Lookup<Input, Key>['type'], Lookup<Data, Key>>
  : IncludesRecord<Lookup<Input, Key>> extends true
  ? RecordSpec<Lookup<Input, Key>, Lookup<Data, Key>>
  : IncludesPlainObjectOrInterface<Lookup<Input, Key>> extends true
  ? NamespaceSpec<Lookup<Input, Key>, Lookup<Data, Key, NO_DATA_MATCH>, LocalContext>
  : LeafSpec<Lookup<Input, Key>, Lookup<Data, Key>>

/**
 * Namespace Utils
 */

export type ExcludeShorthand<T> = Exclude<OnlyPlainObjectOrInterface<T>, Leaf>

type NormalizeNamespaceInput<Input, Data> = NormalizeNamespaceInputFieldsOptionalityWithDataFieldsOptionality<
  ExcludeShorthand<Input>,
  Data
>

type NormalizeNamespaceInputFieldsOptionalityWithDataFieldsOptionality<Input, Data> =
  // input optional + data optional = input optional
  // input optional + data required = input required
  // input optional + no data       = input required
  // input required + *             = input required
  // prettier-ignore
  (InputKeysWhereOptionalAndInDataToo<Input,Data> extends never ? {} : Pick<Input, InputKeysWhereOptionalAndInDataToo<Input,Data>>) &
  (Required<Omit<Input, InputKeysWhereOptionalAndInDataToo<Input,Data>>>)

type InputKeysWhereOptionalAndInDataToo<Input, Data> = {
  [K in keyof Input]: K extends keyof Data
    ? undefined extends Input[K]
      ? undefined extends Data[K]
        ? K
        : never
      : never
    : never
}[keyof Input]

export type GetNamespaceDataWhereDivergingFromInput<Input, Data> = Pick<
  ExcludeUndefined<Data>,
  GetNamespaceDataKeysWhereDivergingFromInput<
    ExcludeShorthand<ExcludeUndefined<Input>>,
    ExcludeUndefined<Data>
  >
>

export type GetNamespaceDataKeysWhereDivergingFromInput<Input, Data> = {
  [K in keyof Data]: K extends keyof Input
    ? IsNamespaceInputFieldEqualDataField<Data[K], Input[K]> extends true
      ? never
      : K
    : K
}[keyof Data]

// optionality differences does not count in determinining diverging
type IsNamespaceInputFieldEqualDataField<InputField, DataField> = IsEqual<
  ExcludeUndefined<InputField>,
  ExcludeUndefined<DataField>
>

export type OmitNamespaceKeys<T> = {
  [K in keyof T]: IsRecord<ExcludeUndefined<T[K]>> extends true
    ? never
    : HasNamespace<ExcludeUndefined<T[K]>> extends true
    ? never
    : T[K]
}

type HasNamespace<T> = OnlyNamespace<T> extends never ? false : true

export type IsNamespaceInputEqualData<Input, Data> = IsEqual<
  Required<OmitNamespaceKeys<UnwrapSyntheticLeavesDeeply<ExcludeShorthand<Input>>>>,
  Required<OmitNamespaceKeys<ExcludeUndefined<Data>>>
>

type OnlyNamespace<T> = T extends Function
  ? never
  : T extends RegExp
  ? never
  : T extends Date
  ? never
  : T extends any[]
  ? never
  : T extends Leaf
  ? never
  : T extends Primitive
  ? never
  : IsRecord<T> extends true
  ? never
  : T

type OnlyShorthand<T> = UnwrapSyntheticLeaf<ExcludeUndefined<Exclude<T, OnlyNamespace<T>>>>

//prettier-ignore
export type NamespaceSpec<Input, Data, LocalContext> =
  {
    fields: NamespaceFields<ExcludeShorthand<Input>, Data>
  } &
  // todo jsdoc 1) when namespace has no matching data key then developer is responsible
  // todo to map the data tree over _somehow_. Impossible to know how, so return type is
  // todo any possible data. This logic is arbitrary and not guaranteed to work. You should
  // todo unit test it!
  (
    Data extends NO_DATA_MATCH
    ? {}  // something already diverged higher up in the hierarchy, then nothing to do here
    : IsNamespaceInputEqualData<Input, Data> extends true
    // ? { i: Input, d: Data, eq: IsNamespaceInputEqualData<Input, Data> }
    ? {} 
    : { map(normalizedInput: NormalizeNamespaceInput<Input, ExcludeUndefined<Data>>, context: BaseContext & LocalContext): GetNamespaceDataWhereDivergingFromInput<Input, Data> }
  ) &
  /**
   * If namespace is union with non-pojo type then shorthand required 
   */
  (
    OnlyShorthand<Input> extends never
    ? {  }
    : {
        shorthand: Shorthand<OnlyShorthand<Input>, UnwrapSyntheticLeavesDeeply<ExcludeShorthand<Input>>>
      }
  ) &
  /**
   * If namespace is optional AND 1+ sub inputs are required THEN initializer required 
   *  ... if no data match THEN still required
   *  ... but if data optional THEN initial is forbidden (since we'll initialize namespace (data) to undefined)
   *  ... but if all namespace fields (input) are optional THEN initial is forbidden (b/c we can automate
   *      namespace (data) with namespace (input) field initializers)
   */
  (
    undefined extends Input
      ? {} extends KeepOptionalKeys<Input>
        ? {}
        : undefined extends Data
          ? { initial?(): KeepRequiredKeys<Exclude<Input, undefined>> }
          : { initial():  KeepRequiredKeys<Exclude<Input, undefined>> }
      : {}
  )

/**
 * todo
 */
// prettier-ignore
export type NamespaceFields<Input, Data> =
  (
    KeysWhereNotPresentOrOptionalInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereNotPresentOrOptionalInData<Input, Data>]+?: Node<Input,Data,Key, {}> }
  ) &
  (
    KeysWhereNotPresentOrRequiredInData<Input, Data> extends never
      ? {}
      : { [Key in KeysWhereNotPresentOrRequiredInData<Input, Data>]-?: Node<Input,Data,Key, {}> }
  )

/**
 * todo: currently assumes Record<string, object>
 */
// prettier-ignore
// todo how does no data match affect this?
export type RecordSpec<Input, Data> =
  (
    {
      /**
       * Specify the settings input for each entry in the record.
       */
      // todo how does no data match affect this?
      entry: Node<
        OnlyPlainObjectOrInterface<ExcludeUndefined<Input>>,
        Data,
        string,
        {
          /**
           * The name under which this entry shows up in the reocrd.
           */
          key: string
        }
      >
    }
  ) &
    /**
     * if input is optional then initial is required
     * unless all
     */
    (
      undefined extends Input
      ? {
          /**
           * Initialize the record with some entries. By default the record will be an empty object.
           */
          initial?(): ExcludeUndefined<Input>
        }
      : {
          /**
           * Initialize the record with some entries. Although you require users to input a record, your initializer will still be run too, if provided.
           */
          initial?(): ExcludeUndefined<Input>
        }
    )

// [1]
// If the field can be undefined it means that initial is not required.
// In most cases it probably means initial won't be supplied. However
// there are may be some odd cases where iniital is present but can
// return undefined.
// prettier-ignore
export type LeafSpec<Input, Data> =
  {
    validate?: Validate<ExcludeUndefined<Input>>
    /**
     * Specify a fixup for this setting.
     *
     * A "fixup" corrects minor problems in a
     * given setting. It also provides a human readable message about what was
     * done and why.
     *
     * Return null if no fixup was needed. Return a fixup object
     * otherwise. The new value should be returned along with a list of one or
     * more messages, one for each thing that was fixed.
     */
    fixup?: Fixup<ExcludeUndefined<Input>>
  } &
  /**
   * if input is optional then initial is required
   * if input is optional and no matching data key then initial is required
   * if input is optional and matching data key optional then initial is optional
   */
  (
    undefined extends Input
      ? undefined extends Data
        ? { initial?(): Input }
        : { initial(): ExcludeUndefined<Input> }
      : {}
  )

/**
 * Isolated Types
 */

export type Validate<Input = Primitive> = (value: Input) => null | { reasons: string[] }

//todo support no messages
export type Fixup<Input = Primitive> = (input: Input) => null | { value: Input; messages?: string[] }

export type MapType<Input = Primitive, Return = Primitive> = (input: Input) => Return

export type Shorthand<Input = Primitive, Return = PlainObject> = (input: Input) => Return

/**
 * Synthetic Leaves
 */

export type Leaf<T = any> = { __settingKind: 'Leaf'; type: T }

export type AlreadyNativeTypeError = 'Error: You wrapped Leaf<> around this field type but Setset already considers it a leaf.'

export type NativeLeaf = Primitive | Date | RegExp

export type UnwrapSyntheticLeavesDeeply<T, CheckUselessWraps extends boolean = false> = {
  [K in keyof T]: T[K] extends Leaf
    ? UnwrapSyntheticLeaf<T[K], CheckUselessWraps>
    : HasNamespace<T[K]> extends true
    ? UnwrapSyntheticLeavesDeeply<T[K]>
    : T[K]
}

export type UnwrapSyntheticLeaf<T, CheckUselessWraps extends boolean = false> = T extends Leaf
  ? CheckUselessWraps extends true
    ? T['type'] extends NativeLeaf
      ? AlreadyNativeTypeError
      : T['type']
    : T['type']
  : T
