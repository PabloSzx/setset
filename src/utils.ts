import { Primitive } from 'type-fest'

/**
 * Check if curerntly in production mode defined as
 * NODE_ENV environment variable equaling "production".
 */
export function isProduction(): boolean {
  return process.env.NODE_ENV === 'production'
}

/**
 * Check if curerntly in development mode defined as
 * NODE_ENV environment variable not equaling "production".
 */
export function isDevelopment(): boolean {
  return process.env.NODE_ENV !== 'production'
}

export type Lookup<Object, Key, Fallback = never> = Key extends keyof Object ? Object[Key] : Fallback

export type IsSameKeys<A, B> = keyof A extends keyof B ? (keyof B extends keyof A ? true : false) : false

export type AnyObject = { [k: string]: any }

export declare type ExcludeUndefined<A> = A extends undefined ? never : A

export type IncludesVoid<T> = (T extends void ? true : never) extends never ? false : true

export type IncludesUndefined<T> = (T extends undefined ? true : never) extends never ? false : true

export type IsRecord<T> = string extends keyof T ? true : false

export type OnlyIndexedType<T> = string extends keyof T ? T : never

export type IncludesRecord<T> = OnlyIndexedType<Exclude<T, Primitive | void>> extends never ? false : true

export type MaybePromise<T = void> = T | Promise<T>

export type CallbackRegistrer<F> = (f: F) => void

export type SideEffector = () => MaybePromise

export type OnlyPlainObjectOrInterface<T> = T extends Function
  ? never
  : T extends RegExp
  ? never
  : T extends Date
  ? never
  : T extends any[]
  ? never
  : T extends Primitive
  ? never
  : T

export type IncludesPlainObjectOrInterface<T> = OnlyPlainObjectOrInterface<T> extends never ? false : true

export type ExcludePlainObjectOrInterface<T> = Exclude<T, OnlyPlainObjectOrInterface<T>>

/**
 * Represents a POJO. Prevents from allowing arrays and functions
 *
 * @remarks
 *
 * TypeScript interfaces will not be considered sub-types. For that, use AnyObject
 */
export type PlainObject = {
  [x: string]: Primitive | object
}

export type isPlainObject<T> = T extends PlainObject ? true : false

export type NotPlainObject = Primitive | any[] | Function

export type IncludesPlainObject<T> = Only<T, PlainObject> extends never ? false : true

export type Only<T, U> = Exclude<T, Exclude<T, U>>

export type KeepOptionalKeys<t> = {
  [k in keyof t]: undefined extends t[k] ? t[k] : never
}

export type GetRequiredKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T]

export type KeepRequiredKeys<t> = {
  [k in ExcludeUndefined<GetRequiredKeys<t>>]-?: undefined extends t[k] ? never : t[k]
}

export type UnknownFallback<T, U> = unknown extends T ? U : T

export type Includes<T, U> = (T extends U ? T : never) extends never ? false : true

export type ExcludePrimitive<T> = Exclude<T, Primitive>

export function assertTypeEquals<T>(expression: T): void {}

export function mergeShallow(o1: any, o2: any) {
  for (const [k, v] of Object.entries(o2)) {
    if (v !== undefined) {
      o1[k] = v
    }
  }
  return o1
}
