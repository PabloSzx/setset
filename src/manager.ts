import * as Logger from '@nexus/logger'
import { forEach, values } from 'lodash'
import { InferDataFromInput, MetadataState } from '.'
import {
  commitNamespace,
  createInfo,
  dataFromMetadata,
  FixupInfo,
  initialize,
  isNamespaceSpecifier,
  isRecordSpecifier,
  normalize,
} from './settings'
import { validateSpecifier } from './spec-validation'
import { Spec, UserInput } from './static'
import { isDevelopment, PlainObject } from './utils'

const log = Logger.log.child('settings')

/**
 * todo
 */
export type Manager<Input extends PlainObject, Data extends PlainObject> = {
  reset(): Manager<Input, Data>
  change(input: UserInput<Input>): Manager<Input, Data>
  original(): Data
  metadata: MetadataState<Data>
  data: Data
}

export type Options = {
  /**
   * Handle fixup events.
   *
   * If your settings spec has no fixups then you can ignore this option.
   *
   * By default, fixups are logged at warning level. If you provide your own
   * function then this default behaviour will be disabled. You can retain it by
   * calling the default function passed as a second argument to your function.
   */
  onFixup?: (info: FixupInfo, originalHandler: (info: FixupInfo) => void) => void
}

function treeifySpec(dec: any, parent: any) {
  dec.parent = parent
  if (isNamespaceSpecifier(dec)) {
    forEach(values(dec.fields), (v: any) => {
      treeifySpec(v, dec)
    })
    return dec
  }
  if (isRecordSpecifier(dec)) {
    treeifySpec(dec.entry, dec)
    return dec
  }
  return dec
}

/**
 *
 */
export function create<Input extends PlainObject, Data extends PlainObject = InferDataFromInput<Input>>(
  specDec: Spec<Input, Data> & Options
): Manager<Input, Data> {
  log.debug('construct')
  const info = createInfo()

  const spec = treeifySpec(specDec, '__ROOT__')

  if (isDevelopment()) {
    validateSpecifier(spec, info)
  }

  // todo we currently have to clone the given spec deeply because mapEntryData mutations the spec with shadow specifiers
  // and shodow specifiers currently break the second+ initialize run (e.g. during reset)

  const initial = initialize(spec, info)
  const state = {
    data: initial.data as Data,
    original: (undefined as any) as Data, // lazy
    metadata: initial.metadata as any, // Metadata<Data>,
  }

  const api: Manager<Input, Data> = {
    data: state.data,
    metadata: state.metadata,
    change(input) {
      log.debug('change', { input })
      const newData = normalize(spec, 'change', spec, input, state.data, state.metadata, info)
      commitNamespace(spec, 'change', newData, state.data, state.metadata, info)
      return api
    },
    reset() {
      log.debug('reset')
      const initial = initialize(spec, info)
      api.data = state.data = initial.data as any
      api.metadata = state.metadata = initial.metadata as any
      return api
    },
    original() {
      log.debug('get original')
      const original = state.original ?? dataFromMetadata(state.metadata, {})
      return original
    },
  }

  return api
}
