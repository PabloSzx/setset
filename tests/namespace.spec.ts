import * as tst from 'typescript-test-utils'
import * as S from '../src'
import { c, R } from './__helpers'

describe('specifier', () => {
  it('can be omitted when input+data is optional', () => {
    const s2 = S.create<{ a?: { b?: number } }, { a?: { b?: number } }>({ fields: {} })
    s2.change({ a: { b: 2 } })
    expect(s2.data).toEqual({ a: { b: 2 } })
  })
})

describe('Well known classes do not get counted as namespace', () => {
  it('regexp', () => {
    S.create<{ a?: { b?: string | RegExp } }>({ fields: { a: { fields: { b: { initial: () => /a/ } } } } })
  })
  it('Date', () => {
    S.create<{ a?: { b?: string | Date } }>({
      fields: { a: { fields: { b: { initial: () => new Date() } } } },
    })
  })
})

describe('static errors', () => {
  it('a field requires an initializer if the input is optional', () => {
    S.create<{ a?: { b?: 1 } }>({ fields: { a: { fields: { b: { initial: c(1) } } } } })
    // @ts-expect-error
    S.create<{ a?: { b?: 1 } }>({ fields: { a: { fields: { b: {} } } } })
  })
  it('a field cannot have an initializer if the input is required', () => {
    S.create<{ a: { b: 1 } }>({ fields: { a: { fields: { b: {} } } } })
    // @ts-expect-error
    S.create<{ a: { b: 1 } }>({ fields: { a: { fields: { b: { initial: c(1) } } } } })
  })
  it('if the namespace is optional but its fields are required then namespace must have initializer', () => {
    S.create<{ a?: { b: 1 } }>({ fields: { a: { initial: c({ b: 1 }), fields: { b: {} } } } })
    // @ts-expect-error missing a.initial
    S.create<{ a?: { b: 1 } }>({ fields: { a: { fields: { b: {} } } } })
  })
  it('namespace initializer only needs to return required fields', () => {
    S.create<{ a?: { b: 1; c?: 2 } }>({
      fields: { a: { initial: c({ b: 1 }), fields: { b: {}, c: { initial: c(2) } } } },
    })
    // @ts-expect-error missing initializer
    S.create<{ a?: { b: 1; c?: 2 } }>({ fields: { a: { initial: c({ b: 1 }), fields: { b: {}, c: {} } } } })
    // prettier-ignore
    // @ts-expect-error missing required fields in namespace initializer
    S.create<{ a?: { b: 1; c?: 2 } }>({ fields: { a: { initial: c({}), fields: { b: {}, c: { initial: c(2) } } } } })
  })
})

describe('initializer', () => {
  it('statically required when namespace optional', () => {
    // @ts-expect-error
    S.create<{ a?: { b: 1 } }>({ fields: { a: { fields: { b: {} } } } })
    S.create<{ a?: { b: 1 } }>({ fields: { a: { initial: c({ b: 1 }), fields: { b: {} } } } })
  })
  it('statically forbidden when namespace optional, but no required fields within', () => {
    // @ ts-expect-error
    // todo not failing right now because of excess property checks
    S.create<{ a?: { b?: 1 } }>({ fields: { a: { fields: { initial: c({}), b: { initial: c(1) } } } } })
  })
  it('statically optional when namespace optional and has 1+ required fields within, but is also optional in the data', () => {
    S.create<{ a?: { b?: 1 } }, { a?: { b: 1 } }>({ fields: { a: { fields: { b: { initial: c(1) } } } } })
    // prettier-ignore
    // todo initial is not actually allowed here, only looks like it b/c of excess property checks
    S.create<{ a?: { b?: 1 } }, { a?: { b: 1 }}>({ fields: { a: { initial: c({}), fields: { b: { initial: c(1) } } } } })
  })
  it('accepted when namespace optional in input but required in data; it initializes trees under required fields', () => {
    const s = S.create<{ a?: { b: 1 } }>({ fields: { a: { initial: c({ b: 1 }), fields: { b: {} } } } })
    expect(s.data).toEqual({ a: { b: 1 } })
  })
  it('statically (todo) cannot provide optional fields and if it does are overriden by the field initializer', () => {
    // todo this should raise a compiler warning but it does not because of https://github.com/microsoft/TypeScript/issues/241#issuecomment-669138047
    // @ ts-expect-error c prop forbidden from namespace initializer
    // prettier-ignore
    const s = S.create<{ a?: { b: 1, c?: number } }>({ fields: { a: { log:{b:1,c:2}, log2: {b:1,c:2}, initial(){ return { b:1, c: 3 } }, fields: { b: {}, c:{initial: c(2) } } } } })
    expect(s.data).toEqual({ a: { b: 1, c: 2 } })
  })
})

it('a field initializer initializes its data', () => {
  const settings = S.create<{ a?: { b?: 1 } }>({ fields: { a: { fields: { b: { initial: c(1) } } } } })
  expect(settings.data.a.b).toEqual(1)
})
it('data can be changed', () => {
  const settings = S.create<{ a?: { b?: number } }>({ fields: { a: { fields: { b: { initial: c(1) } } } } })
  expect(settings.change({ a: { b: 2 } }).data).toEqual({ a: { b: 2 } })
})
it('data merges deeply preserving existing data not targetted by input', () => {
  const settings = S.create<{ a?: { a?: number; b?: number }; b?: number }>({
    fields: {
      a: { fields: { a: { initial: c(1) }, b: { initial: c(2) } } },
      b: { initial: c(3) },
    },
  })
  expect(settings.change({ a: { a: 4 } }).data).toEqual({ a: { a: 4, b: 2 }, b: 3 })
})
// todo bring back in strict mode, see commented out code in resolve func
it.skip('passing a plain object to a non-namespace field will error gracefully', () => {
  const s = S.create<{ a?: 1 }>({ fields: { a: { initial: c(1) } } })
  // @ts-expect-error
  expect(() => s.change({ a: { b: 2 } })).toThrowErrorMatchingSnapshot()
})
describe('with shorthands', () => {
  it('allow a value to be assigned directly to the namespace field', () => {
    type i = { a?: number | { b?: number } }
    type d = { a: { b: number } }
    const settings = S.create<i, d>({
      // prettier-ignore
      fields: { a: { shorthand: (value) => ({ b: value + 100 }), fields: { b: { initial: c(1) } } } }
    })
    expect(settings.data).toEqual({ a: { b: 1 } })
    expect(settings.change({ a: 2 }).data).toEqual({ a: { b: 102 } })
  })
  describe('runtime errors', () => {
    it('unexpected shorthand errors fail gracefully', () => {
      type d = { a: { b: number } }
      type i = { a: number | { b: number } }
      // prettier-ignore
      const s = S.create<i, d>({ fields: { a: { shorthand(value) { throw new Error(`Unexpected shorthand error with value ${value}`) }, fields: { b: { initial: c('') } } } } })
      expect(() => s.change({ a: 100 }).data.a).toThrowErrorMatchingSnapshot()
    })
  })
  it('still accepts longhand input', () => {
    type d = { a: { b: number } }
    type i = { a: number | { b: number } }
    const s = S.create<i, d>({
      // prettier-ignore
      fields: { a: { shorthand: (n) => ({ b: n + 100 }) , fields: { b: { initial: c(1) } } } }
    })
    expect(s.change({ a: { b: 3 } }).data.a).toEqual({ b: 3 })
  })
  it('can be a type that differs from the longhand', () => {
    const s = S.create<{ a: (() => number) | { b: string } }, { a: { b: string } }>({
      // prettier-ignore
      fields: { a: { shorthand: (f) => ({ b: f().toString() }), fields: { b: {} } } }
    })
    expect(s.change({ a: () => 1 }).data).toEqual({ a: { b: '1' } })
  })

  it('changing with a shorthand on a namespace that does not support them will error gracefully', () => {
    const s = S.create<{ a: { b: string } }>({ fields: { a: { fields: { b: {} } } } })
    // @ts-expect-error
    expect(() => s.change({ a: 'runtime error' })).toThrowError(
      'Setting "a" is a namespace with no shorthand so expects an object but received a non-object: \'runtime error\''
    )
  })
})

describe('map on root namespace', () => {
  it('it works on root', () => {
    // prettier-ignore
    const s = S.create<{ b1: string }, { b2: string }>({
      map(input, ctx) { return { b2:'bar' } },
      fields: { b1: {} }
    })
    expect(s.change({ b1: 'foo' }).data).toEqual({ b1: 'foo', b2: 'bar' })
  })
})

describe('map on record entry namespace', () => {
  it('context param contains "key" prop indicating where in record this namespace comes from', () => {
    // prettier-ignore
    const s = S.create<{ a: R<{ b1: string }> }, { a: R<{ b2: {path:string[],key:string} }> }>({
        fields: { a: { entry: {
            map(input, ctx) { tst.assertTrue<tst.Equals<{path:string[],key:string}, typeof ctx>>(); return {b2:ctx} },
            fields: { b1: {} } }
        }}
      })
    expect(s.change({ a: { foobar: { b1: 'ignore' } } }).data).toEqual({
      a: { foobar: { b1: 'ignore', b2: { path: ['__root__', 'a', 'foobar'], key: 'foobar' } } },
    })
  })
  it('runs over entries provided by record initializer', () => {
    // prettier-ignore
    const s = S.create<{ a: R<{ b1: string }> }, { a: R<{ b2: {input: {b1:string},ctx: {path:string[],key:string}} }> }>({
      fields: { a: {
        initial: c({foo:{b1:''}}),
        entry: {
          map(input, ctx) { return {b2:{input:{...input}, ctx}} },
          fields: { b1: {} }
        }
      }}
    })
    expect(s.data).toMatchSnapshot()
  })
})

describe('map with initializers does not show up in the metadata data, and...', () => {
  it('runs at initialization time with input from namesapce initializer', () => {
    // prettier-ignore
    const s = S.create<{ a: { b?: { c1:string }  } }, { a: { b: { c2: string} } }>({
        fields: { a: { fields: { b: {
          initial(){ return { c1:'foo'} },
          map(input, ctx) { return { c2: input.c1 } },
          fields: { c1: {} }
        }}}}
      })
    expect(s.data).toEqual({ a: { b: { c1: 'foo', c2: 'foo' } } })
    expect(s.metadata).toMatchSnapshot()
  })
  it('runs at initialization time with input from namesapce field initializers', () => {
    // prettier-ignore
    const s = S.create<{ a: { b1?: string } }, { a: { b2: string } }>({
        fields: { a: {
          map(input, ctx) { return {b2:input.b1} },
          fields: { b1: { initial: c('foo')} }
        }}
      })
    expect(s.data).toEqual({ a: { b1: 'foo', b2: 'foo' } })
    expect(s.metadata).toMatchSnapshot()
  })
  it('runs at initialization time with input from namespace and field initializers', () => {
    // prettier-ignore
    const s = S.create<{ a: { b?: { c1:string, d1?:string }  } }, { a: { b: { c2: string, d2: string} } }>({
        fields: { a: { fields: { b: {
          initial(){ return { c1:'c1-foo'} },
          map(input, ctx) { return { c2: input.c1, d2: input.d1 } },
          fields: { c1: {}, d1: { initial: c('d1-foo') } }
        }}}}
      })
    expect(s.data).toEqual({ a: { b: { c1: 'c1-foo', c2: 'c1-foo', d1: 'd1-foo', d2: 'd1-foo' } } })
    expect(s.metadata).toMatchSnapshot()
  })
})

describe('map on non-root namespace', () => {
  it.todo('gracefully errors when unexpected error')
  it('when input is not in data it is statically unavailable in data but actually present there at runtime', () => {
    // prettier-ignore
    const s = S.create<{ a: { b1: string } }, { a: { b2: string } }>({
        fields: { a: {
          map(input, ctx) { return {b2:'bar'} },
          fields: { b1: {} }
        }}
      })
    expect(s.change({ a: { b1: 'foo' } }).data).toEqual({ a: { b1: 'foo', b2: 'bar' } })
  })
  it('input param (1st) is the normalized input (e.g. shorthands are expanded)', () => {
    const map = jest.fn().mockImplementation((input, ctx) => {
      return { b: Number(input.b) }
    })

    // prettier-ignore
    const s = S.create<{ a: string | { b: string } }, { a: { b: number } }>({
      fields: { a: {
        map,
        shorthand(s) { return {b:s} },
        fields: { b: {} } } }
    })
    expect(s.change({ a: '1' }).data).toEqual({ a: { b: 1 } })
    expect(map.mock.calls).toMatchSnapshot()
  })
  it('context param (2nd) contains path prop', () => {
    // prettier-ignore
    const s = S.create<{ a: { b: string } }, { a: { b: {path:string[]} } }>({
        fields: { a: {
          map(input, ctx) { tst.assertTrue<tst.Equals<{path:string[]}, typeof ctx>>(); return {b:ctx} },
          fields: { b: {} } } }
      })
    expect(s.change({ a: { b: 'ignore' } }).data).toEqual({ a: { b: { path: ['__root__', 'a'] } } })
  })
})
