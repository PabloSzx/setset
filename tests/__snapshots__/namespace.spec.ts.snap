// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map on non-root namespace input param (1st) is the normalized input (e.g. shorthands are expanded) 1`] = `
Array [
  Array [
    Object {
      "b": 1,
    },
    Object {
      "path": Array [
        "__root__",
        "a",
      ],
    },
  ],
  Array [
    Object {
      "b": "1",
    },
    Object {
      "path": Array [
        "__root__",
        "a",
      ],
    },
  ],
]
`;

exports[`map on record entry namespace runs over entries provided by record initializer 1`] = `
Object {
  "a": Object {
    "foo": Object {
      "b1": "",
      "b2": Object {
        "ctx": Object {
          "key": "foo",
          "path": Array [
            "__root__",
            "a",
            "foo",
          ],
        },
        "input": Object {
          "b1": "",
        },
      },
    },
  },
}
`;

exports[`map with initializers does not show up in the metadata data, and... runs at initialization time with input from namesapce field initializers 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "fields": Object {
        "b1": Object {
          "from": "initial",
          "initial": "foo",
          "type": "leaf",
          "value": "foo",
        },
      },
      "type": "namespace",
    },
  },
  "type": "namespace",
}
`;

exports[`map with initializers does not show up in the metadata data, and... runs at initialization time with input from namesapce initializer 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "fields": Object {
        "b": Object {
          "fields": Object {
            "c1": Object {
              "from": "initial",
              "initial": "foo",
              "type": "leaf",
              "value": "foo",
            },
          },
          "type": "namespace",
        },
      },
      "type": "namespace",
    },
  },
  "type": "namespace",
}
`;

exports[`map with initializers does not show up in the metadata data, and... runs at initialization time with input from namespace and field initializers 1`] = `
Object {
  "fields": Object {
    "a": Object {
      "fields": Object {
        "b": Object {
          "fields": Object {
            "c1": Object {
              "from": "initial",
              "initial": "c1-foo",
              "type": "leaf",
              "value": "c1-foo",
            },
            "d1": Object {
              "from": "initial",
              "initial": "d1-foo",
              "type": "leaf",
              "value": "d1-foo",
            },
          },
          "type": "namespace",
        },
      },
      "type": "namespace",
    },
  },
  "type": "namespace",
}
`;

exports[`with shorthands runtime errors unexpected shorthand errors fail gracefully 1`] = `
"There was an unexpected error while running the namespace shorthand for setting \\"a\\". The given value was 100 
Unexpected shorthand error with value 100"
`;
